cpu_freq: Add CPU frequency scaling subsystem

seankyer opened the pull request:
This PR implements: https://github.com/zephyrproject-rtos/zephyr/issues/92123 CPU Frequency Scaling Subsystem

**Updated System Level Diagram**
<img width="766" height="455" alt="image" src="https://github.com/user-attachments/assets/839492d7-c18b-4bc3-97fd-edccfe9d50df" />

**Differences from Original RFC Proposal**
- During development it became clear that a P-state policy is very coupled to a given metric (Metric in this case refers to CPU load, CPU temperature, or some other statistic that CPU Freq decisions could be based on). 
  - Based on this, we tweaked the design to create a `metrics` folder within the subsystem of which different metrics could be defined
  - A standard (or custom) policy can then utilize a metric (or set of metrics) to make P-state decisions.
    - The policy itself would make direct calls to the metric
- We did not refactor or promote any existing subsystem to be a standalone CPU frequency subsystem
  - `THREAD_RUNTIME_STATS` option is a lightweight way to get this information and essentially is the "subsystem" we had originally hoped for.

**Future Improvements and Features**
- The current implementation of CPU freq subsystem does not support SMP
  - Tentative solution: worker thread is updated to iterate over all cores and cpu_load metric updated to take cpu core ID as argument

seankyer committed:
cpu_freq: Add CPU Frequency Scaling subsystem

Add a CPU frequency scaling subsystem, allowing a policy
algorithm to control the frequency of a given SoC/MCU
automatically at runtime.

Implement a basic, "on-demand" policy algorithm which
iterates through the P-states supported by the SoC and
selects the first P-state where it's trigger threshold is
less than the CPU load.

The CPU frequency scaling subsystem does not currently
support SMP. The CPU load measurement can be made to support
SMP since statistics are measured from the scheduler.

Co-authored-by: Eric Hay <Eric.Hay@analog.com>
Signed-off-by: Sean Kyer <Sean.Kyer@analog.com>

seankyer commented:
Is someone able to link this PR to https://github.com/zephyrproject-rtos/zephyr/issues/92123 ? I don't seem to have the permission for it.

andyross commented on subsys/cpu_freq/cpu_freq.c:
```diff
@@ -0,0 +1,57 @@
+LOG_MODULE_REGISTER(cpu_freq, CONFIG_CPU_FREQ_LOG_LEVEL);
+
+static void cpu_freq_work_handler(struct k_work *work);
+K_WORK_DELAYABLE_DEFINE(cpu_freq_work, cpu_freq_work_handler);
```
Can this be a k_timer and not a work item?  The former is a lightweight callback that runs in the timer ISR, the latter demands that the system work queue be instantiated, which is a whole thread.  Small apps don't necessarily want to pay a cost like that, and my gut says it's not needed here.  Is there a reason you need the governor code to run in a thread?

Also note that this introduces a priority inversion case: if you have a very high priority thread spinning, the work queue might not run, and so might not be able to boost the CPU speed for the very high priority spinning thread!

seankyer replied:
Good idea, I switched it to use a timer instead.

andyross reviewed (commented):
Notes & nitpicks.  Nothing seems too scary from my perspective.  The big points to note are: "We need a story for how this API extends to multiprocessor setups" and "The governor code probably wants to run in a timer and not the work queue"

ehay64 commented on subsys/cpu_freq/metrics/cpu_load.c:
```diff
@@ -0,0 +1,40 @@
+static uint64_t execution_cycles_prev;
+static uint64_t total_cycles_prev;
+
+int get_cpu_load(uint32_t *load)
```
There's already a few different API's available to retrieve this type of information, it really depends on the use case:
* subsys/debug/cpu_load.c
* subsys/debug/thread_analyzer
* Object Core Statistics

I figured it would be best to rely on an already established API to retrieve these stats. The call we use, `k_thread_runtime_stats_cpu_get()`, and the infrastructure needed for it is also pretty lightweight. Also, this implementation is fairly application specific. It relies on stable periodic calls to produce consistent results. I think if we want to promote this type of functionality, it would require another discussion.

andyross commented on include/zephyr/cpu_freq/p_state.h:
```diff
@@ -0,0 +1,49 @@
+struct p_state {
+	uint32_t load_threshold; /**< CPU load threshold at which P-state should be triggered */
+	uint8_t disabled;        /**< Flag to indicate if P-state is disabled */
+	const void *config;      /**< Vendor specific devicetree properties */
```
The pstate "owner" could define:

```c
struct my_pstate {
    struct p_state pstate;
    int my_config_data;
};
```

And then any time it wants to get access to my_config_data from the callback, it can use CONTAINER_OF().  Now the struct doesn't need the pointer in it anymore.

bjarki-andreasen commented on include/zephyr/cpu_freq/p_state.h:
```diff
@@ -0,0 +1,49 @@
+ */
+#define P_STATE_DT_GET(_node) &(_CONCAT(p_state_, DEVICE_DT_NAME_GET(_node))),
+
+struct p_state {
```
These states have to be in the devicetree, they are entirely hardware specific, they describe combinations of voltages and clock speeds supported by the hardware, something like clock states or pinctrl states, or indeed sleep states.

seankyer replied:
so just to clarify, as I have it, is correct?

decsny commented on samples/subsys/cpu_freq/on_demand/src/main.c:
```diff
@@ -0,0 +1,67 @@
+	/* Set timer to change cpu scenario periodically */
+	k_timer_start(&timer, K_MSEC(10000), K_MSEC(10000));
+
+	while (1) {
+		switch (cpu_scenario) {
+		case CPU_SCENARIO_0:
+			/* CPU is always in sleep */
+			k_usleep(100000);
+			break;
+		case CPU_SCENARIO_1:
+			/* CPU doing some intermittent processing */
+			for (int i = 0; i < 1000; i++) {
+				counter++;
+			}
+			k_usleep(200);
+			break;
+		case CPU_SCENARIO_2:
+			/* CPU doing some more intense processing */
+			for (int i = 0; i < 1000; i++) {
+				counter++;
+			}
+			k_usleep(1);
+			break;
+		case CPU_SCENARIO_4:
+			/* CPU doing lots of calculations */
+			counter++;
+			break;
+		default:
+			LOG_ERR("Unknown CPU scenario: %d", cpu_scenario);
+			return -1;
+		}
+	}
```
some logging in the sample might be more interesting

seankyer replied:
The example prj.conf configures the subsystem DBG logs to output a lot of information:
<img width="1331" height="799" alt="image" src="https://github.com/user-attachments/assets/c8651f21-258e-46c3-873c-380bdb7cd4c1" />

Which I think is pretty verbose already.

bjarki-andreasen reviewed (changes requested):
Please review and apply the changes in these commits https://github.com/zephyrproject-rtos/zephyr/pull/93905/commits/d027ff6d2a5cb61af92fcc663fc84e984801ea48 and https://github.com/zephyrproject-rtos/zephyr/pull/93905/commits/0d38eba1119b728dc19a673b5afbde5dc602cee8 from #93905 :)

seankyer commented:
> Please review and apply the changes in these commits [d027ff6](https://github.com/zephyrproject-rtos/zephyr/commit/d027ff6d2a5cb61af92fcc663fc84e984801ea48) and [0d38eba](https://github.com/zephyrproject-rtos/zephyr/commit/0d38eba1119b728dc19a673b5afbde5dc602cee8) from #93905 :)

I applied the changes to the original commit. Thanks for helping make things more "Zephyr-like". 

I did not apply the change to remove the 'disabled' prop, as I want to leave it open for extension to introduce an API allowing application to disable a pstate at runtime.

seankyer commented:
> we need tests, samples are good, but not enough, please add tests for this new subsystem.

Okay, right. I added two tests. One to test SoC integration and another to test the policy (for each policy a new test should be added).

seankyer committed:
tests: cpu_freq: Add SoC and on_demand tests

Add tests for CPU frequency subsystem covering
SoC integration of pstate as well as the on_demand
policy.

Signed-off-by: Sean Kyer <Sean.Kyer@analog.com>

bjarki-andreasen commented:
> > Please review and apply the changes in these commits [d027ff6](https://github.com/zephyrproject-rtos/zephyr/commit/d027ff6d2a5cb61af92fcc663fc84e984801ea48) and [0d38eba](https://github.com/zephyrproject-rtos/zephyr/commit/0d38eba1119b728dc19a673b5afbde5dc602cee8) from #93905 :)
> 
> I applied the changes to the original commit. Thanks for helping make things more "Zephyr-like". 

Cool :)

> I did not apply the change to remove the 'disabled' prop, as I want to leave it open for extension to introduce an API allowing application to disable a pstate at runtime.

The disabled prop is in the devicetree, so it can't be changed at runtime?

seankyer commented:
> The disabled prop is in the devicetree, so it can't be changed at runtime?

You're right, I got confused. Take a look now, thanks

bjarki-andreasen commented on samples/subsys/cpu_freq/on_demand/src/main.c:
```diff
@@ -0,0 +1,67 @@
+			break;
+		case CPU_SCENARIO_1:
+			/* CPU doing some intermittent processing */
+			for (int i = 0; i < 1000; i++) {
```
Use k_busy_wait() to spin the CPU for a period

seankyer replied:
thanks. I refactored the example slightly and I think its clearer now.

bjarki-andreasen commented on include/zephyr/cpu_freq/pstate.h:
```diff
@@ -0,0 +1,59 @@
+#define PSTATE_DT_DEFINE(_node, _config)                                                           \
+	const struct pstate PSTATE_DT_SYM(_node) = {                                               \
+		.load_threshold = DT_PROP(_node, load_threshold),                                  \
+		.disabled = false,                                                                 \
```
Given this struct is const, this member is redundant (its basically a #define 0)

bjarki-andreasen commented on include/zephyr/cpu_freq/policy.h:
```diff
@@ -0,0 +1,37 @@
+ *
+ * @retval 0 In case of success, nonzero in case of failure.
+ */
+int cpu_freq_policy_select_pstate(struct pstate *pstate);
```
When passing pstates around, given they are const, and larger than a pointer, use pointers:
```
int cpu_freq_policy_select_pstate(const struct pstate **pstate);
```
The additional cost of copying is typically only accepted for syscalls, where data may not be accessible to the caller

bjarki-andreasen reviewed (changes requested):
Looking good! A few last nits

seankyer commented:
cherry-picked some fixes for max32655 in the meantime: https://github.com/zephyrproject-rtos/zephyr/pull/94032

bjarki-andreasen commented:
The last two commits seem unrelated? otherwise looks good

nashif commented on subsys/cpu_freq/metrics/cpu_load.c:
```diff
@@ -0,0 +1,40 @@
+static uint64_t execution_cycles_prev;
+static uint64_t total_cycles_prev;
+
+int cpu_freq_metric_get_cpu_load(uint32_t *load)
```
This should be split out and provided on its own, does not need to be part of this subsystem, this way it can be used by others. Same suggestion was made here: https://github.com/zephyrproject-rtos/zephyr/pull/93461#discussion_r2226899062

I just tried moving this out and it works nicely on its own and without dependency on cpu_freq subsystem. We do not want yet another cpu_load utility hidden under a subsystem. The one we have in subsys/debug probably should be folded into this as this is multi-arch and will support multi-core eventually.

Maybe move this to lib/os or kernel/

nashif replied:
also, naming is inconsistent, you have for example `cpu_freq_pstate_set`, but here you have `cpu_freq_metric_get_cpu_load`,  you should move the `get` to the end.

nashif commented on subsys/cpu_freq/Kconfig:
```diff
@@ -0,0 +1,55 @@
+
+config CPU_FREQ_METRIC_CPU_LOAD
+	bool "Measure CPU Load for Frequency Policies"
+	default n
```
`default n` is not needed

nashif commented on doc/services/cpu_freq/metrics/cpu_load.rst:
```diff
@@ -0,0 +1,15 @@
+CPU Load Metric
+###############
+
+The CPU Load metric makes use of :ref:`thread_analyzer` to implement, ``get_cpu_load(uint32_t *load)``
```
I do not think this is accurate, you do not use thread analyzer, but the `usage` kernel feature, which is also used by thread analyzer.

seankyer replied:
right, I updated the wording.

MaureenHelm reviewed (commented):
Overall looks great, but a question about the hardware implementation. If the `cpu_freq_pstate_set` is a generic mock implementation, maybe better to do on `native_sim` instead of max32?

seankyer committed:
native: cpu_freq: Add CPU freq support to native_sim

Define P-states for native_sim and add mock cpu_freq
driver.

Signed-off-by: Sean Kyer <Sean.Kyer@analog.com>

ehay64 committed:
cpu_load: Add CPU Load subsystem

Implement CPU load measurement subsystem. The statistics
used to calcualte the load are taken from the scheduler.
Specifically, the number of cycles used by the idle thread.

Co-authored-by: Sean Kyer <Sean.Kyer@analog.com>
Signed-off-by: Eric Hay <Eric.Hay@analog.com>

seankyer commented:
The CI failure is unrelated but I've pushed a PR to fix it: https://github.com/zephyrproject-rtos/zephyr/pull/94479 will rebase once merged

seankyer commented:
Latest update:
@MaureenHelm I replaced the max32 pstate dummy implementation with `native_sim` as you suggested
@nashif As you suggested, we moved the `cpu_load` metric out of `cpu_freq` and updated the implementation where relevant

Please take a look when available and let me know. Hoping to rebase soon once https://github.com/zephyrproject-rtos/zephyr/pull/94479 is fixed.

decsny reviewed (changes requested):
I am not sure if this is thought out enough. Your example is using the native sim for which changing the clock is not even a real thing, it just prints a debug message. In a real system to change the frequency and voltage could be a very complicated thing, possibly needing to call on a lot of different software components. Such as:
- Clock manager to reconfigure the whole clock tree of the SOC if changing the cpu clock affects other active clocks in the system, and having to weigh it's own constraints and notify all the affected devices
- Potentially multiple regulators needed to change the voltage of the system 
- Potentially some of these actions to change regulators and clocks actually needing to communicate on an external bus, even a slow and unreliable serial bus such as I2C

So there can be all these latencies and things affected, and on top of that, we don't know what is the future load, ie what is the load pattern, beyond just an average. I think what you have done is provide the possibility of some backends for policies, but the current one is such an oversimplification of the real life problem.

It's so easy to have problems where, if the period is configured by the kconfig to be too long, then you have an issue where the nothing happens for that period, it sets the frequency way too low and all the sudden tons of things start happening on the next period, or vice versa. And if you try to fix this by making the period shorter, you run into the real life problem of all those latencies I mentioned above, which might end up taking up most of the period to actually do, and leaving little time for real processing, and maybe even affecting this algorithm by influencing what the cpu load is, and now you are just optimizing the cpu frequency for running all this potential io/compute to actually do the DVFS.  Imagine, it changes it one cycle to be higher frequency, requiring a bunch of io/processing to reconfigure the system, it thinks its good for one period maybe, then it doesnt do any more changing, now the next time since it didnt do any more changing of the frequency, it thinks it has low load, so it reconfigures everything again, then next period, since it did all that reconfiguring, it wants it to be high again, etc etc on repeat. And like I said, if the period is to long, then it will possibly not be a correctly functioning embedded system, because there are real time requirements and some bottleneck for waiting for the next cpu freq scheduling point can cause some real problem. This amount of non-determinism or unresponsiveness is needing to have some serious thought to minimize, and having this dud of an implementation right now is not convincing to say this should be merged.

IMO we should follow the linux kernel rule of thumb of proposing apis: one implementation means the proposal will likely not work and require massive rework when another thing comes along, two implementations means maybe the proposal is gaining some noteworthy traction but still be cautions, and three implementations, across different hardware architectures, is the golden rule of thumb to have confidence a thing will work.

IF some people want to see this done, then there needs to be an effort by multiple hardware vendors with different enough architectures to actually implement a real life thing and show some actual sample applications doing something besides a busy wait, otherwise we risk this to just become dead code when we find out we have to either do a major rework of it or start from scratch later when the real use cases come knocking about the problems and limitations.

I also want to see some thought put into how would this integrate with a clock managing scheme which is being talked about sometimes in arch WG. I think this cpufreq/dvfs, power management, and clock manager are all deeply related to one another (dvfs maybe is mostly related to power management only indirectly through the clock manager and the regulators of the system supplies). Right now we have a power management with a good architecture, but the missing link is the clock manager, and this p state thing is starting to overlap with @danieldegrasse proposal of clock states too, I think we might end up with some redundancy. I actually don't like the direction of how the clock managing is going right now, I think combining this discussion about DVFS with it will maybe start giving it a real direction to go with.

decsny commented:
To be clear, I actually really do like the proposal and the straightforwardness/directness/simplicity of it, and I think it is likely the right direction to be going. But I am not comfortable merging something as important as this without seeing it actually used for something besides logging debug messages based on a busy wait in posix "simulator" running on a very powerful non-embedded machine in a program with 0 requirements or functionality at all.

aescolar reviewed (commented):
A bit of feedback around details.
Note that I'm not "requesting changes" so I do not need to re-review to unblock the PR, but please address the issues which are not just nits or questions.

seankyer commented on subsys/cpu_freq/Kconfig:
```diff
@@ -0,0 +1,45 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config HAS_CPU_FREQ
```
I'm inclined to leave it the way it is for now, since CPUFreq is the name of the subsystem itself, which is mirroring the name of the Linux subsystem. I'm open to revisiting later however 😄

bjarki-andreasen replied:
The `HAS` is because I copied `HAS_PM`, though it made sense to keep the same style

kartben commented on doc/services/cpu_freq/index.rst:
```diff
@@ -0,0 +1,66 @@
+.. _cpu_freq:
+
+CPU Frequency Scaling Subsystem
+###############################
```
```suggestion
CPU Frequency Scaling
#####################
```

kartben commented on doc/services/cpu_load/index.rst:
```diff
@@ -0,0 +1,16 @@
+.. _cpu_load_subsys:
+
+CPU Load Subsystem
+##################
```
```suggestion
CPU Load
########
```

kartben commented on include/zephyr/cpu_freq/cpu_freq.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <zephyr/types.h>
+#include <zephyr/cpu_freq/pstate.h>
+
```
your headers need to be organized in doxygen groups (probably one for CPU Frequency Scaling and one for CPU Load, I'm guessing) that are "mounted" into the [`os_services` group](https://docs.zephyrproject.org/latest/doxygen/html/group__os__services.html)

kartben commented on include/zephyr/cpu_freq/pstate.h:
```diff
@@ -0,0 +1,57 @@
+ */
+#define PSTATE_DT_GET(_node) &PSTATE_DT_SYM(_node)
+
+struct pstate {
```
missing documentation for the struct itself

kartben commented on include/zephyr/cpu_load/cpu_load.h:
```diff
@@ -0,0 +1,33 @@
+ * Return the percent that the CPU has spent in the active (non-idle) state between calls to this
+ * function.
+ *
+ * @param load Pointer to the integer where the CPU load as a percent (0-100) is returned.
+ * @param cpu_id The ID of the CPU for which to get the load.
+ *
+ * @retval 0 In case of success, nonzero in case of failure.
+ */
+int cpu_load_get(uint32_t *load, int cpu_id);
```
mostly nits but I would find it more straightforward to have the output parameter appear last, i.e. after cpu_id.

```suggestion
 * @param[out] load Pointer to the integer where the CPU load as a percent (0-100) is returned.
 * @param cpu_id The ID of the CPU for which to get the load.
 *
 * @retval 0 In case of success, nonzero in case of failure.
 */
int cpu_load_get(uint32_t *load, int cpu_id);
```

seankyer replied:
I agree, I swapped the order. Thanks for the review.

kartben commented on samples/subsys/cpu_freq/on_demand/README.rst:
```diff
@@ -0,0 +1,14 @@
+Overview
+********
+
+This sample demonstrates the CPU frequency subsystem's on-demand policy. The
```
cross-linking to the cpu freq main doc page would be a good idea

kartben commented on samples/subsys/cpu_freq/on_demand/src/main.c:
```diff
@@ -0,0 +1,73 @@
+LOG_MODULE_REGISTER(cpu_freq_sample, LOG_LEVEL_INF);
+
+#define MS_TO_US(x)      ((x) * 1000)
+#define PERCENT_SLEEP_MS CONFIG_CPU_FREQ_INTERVAL_MS / 100
```
```suggestion
#define PERCENT_SLEEP_MS (CONFIG_CPU_FREQ_INTERVAL_MS / 100)
```

kartben commented on samples/subsys/cpu_freq/on_demand/README.rst:
```diff
@@ -0,0 +1,14 @@
+.. zephyr:code-sample:: cpu_freq_on_demand
+   :name: CPU Frequency Subsystem On-demand Sample
+
+   Demonstrates On-demand CPU frequency scaling policy.
```
minor tweaks to align better with recommended style for describing samples

```suggestion
   :name: On-demand CPU frequency scaling

   Dynamically scale CPU frequency using the on-demand policy.
```

seankyer commented on include/zephyr/cpu_freq/cpu_freq.h:
```diff
@@ -0,0 +1,33 @@
+ *
+ * @retval 0 in case of success, nonzero in case of failure.
+ */
+int cpu_freq_pstate_set(const struct pstate *state);
```
I think this comment and https://github.com/zephyrproject-rtos/zephyr/pull/93461#discussion_r2284316863 can be addressed by better documentation on this function. 

I don't want to impose any requirements on the SoC p-state driver, so I would say that it is not a guarantee that the p-state was applied when this `cpu_freq_pstate_set` is called. The p-state driver may respond with some `EBUSY` error code if it hasn't finished the previous p-state transition. Or the driver can simply ignore the call to `_set` and still return the 0-status code.

I will document this intended behavior.

seankyer commented on tests/subsys/cpu_freq/cpu_freq_soc/src/main.c:
```diff
@@ -0,0 +1,40 @@
+		const struct pstate *state = soc_pstates_dt[i];
+
+		/* Set performance state using pstate driver */
+		ret = cpu_freq_pstate_set(state);
```
see https://github.com/zephyrproject-rtos/zephyr/pull/93461#discussion_r2285862713 for response to this note

seankyer commented on tests/subsys/cpu_freq/policies/on_demand/src/main.c:
```diff
@@ -0,0 +1,44 @@
+		      "Expected -EINVAL for NULL pstate_out");
+
+	/* Get pstate at near 100% load */
+	k_busy_wait(10000);
```
I made a #define for it. It's arbitrary but just meant to simulate the CPU doing some work before calling the pstate. I improved this test a little bit. Thanks.

bjarki-andreasen replied:
k_busy_wait is in microseconds

aescolar replied:
> I made a #define for it

I don't see any change, did you forget to push?

> k_busy_wait() is in microseconds

I don't follow. 10ms is the default system tick period for native_sim.

bjarki-andreasen replied:
> > I made a #define for it
> 
> I don't see any change, did you forget to push?
> 
> > k_busy_wait() is in microseconds
> 
> I don't follow (10ms is the default system tick period for native_sim)

ah, I did not know that, I thought you where asking to what the period actually was, I believe its arbitrary :)

seankyer commented:
@decsny I think you have system-level concerns of this subsystem and a concern about there being no real SoC implementation with this PR.

To address the SoC implementation concern, I should put into context that this RFC has been reviewed by the Architecture WG about a month ago. I've collaborated with community members during the WG, RFC review and this PR to produce the current implementation. There has been engagement across multiple vendors. @bjarki-andreasen has already mocked support for his Nordic SoC [https://github.com/zephyrproject-rtos/zephyr/pull/93905](https://github.com/zephyrproject-rtos/zephyr/pull/93905), and I believe @peter-mitsis at Intel has as well. At ADI we are also implementing support on our SoC which we will upstream.

I believe it's reasonable for this PR to be merged with the dummy `native_sim` driver, serving as both a CI target and a reference implementation, with support for other SoCs to be added through subsequent PRs.

I believe your system concerns are addressable with future development of this subsystem, its policies, and API.

> In a real system to change the frequency and voltage could be a very complicated thing, possibly needing to call on a lot of different software components

Correct. That complexity is real, but it doesn't invalidate having a common API/abstraction layer. Any vendor supporting Zephyr's CPUFreq subsystem must account for those details in their p-state driver and select an appropriate combination of policy and period. If an SoC cannot switch DVFS modes in a performant way, then a load-based policy may not make sense for it (and it should select another).

> we don't know what is the future load, ie what is the load pattern, beyond just an average. I think what you have done is provide the possibility of some backends for policies, but the current one is such an oversimplification of the real life problem.

This concern, and the concern of a circular load-loop, is a fair critique of the _on-demand_ policy, but not of the subsystem itself. If the policy is too simple, the solution is to add a better one. I see there being many policies eventually, some of which combine multiple metrics. Looking ahead, there has already been discussions in the Architecture WG to add a temperature metric and a temperature-based policy https://github.com/zephyrproject-rtos/zephyr/issues/92123#issuecomment-3049414846. I can also think of the rationale behind a load-trend based policy. The API could also be extended with a "force" function, letting an application request a p-state preemptively if it knows heavy processing is coming.

The on-demand policy may indeed be too simple for some systems, but where do we start? It makes perfect sense for the initial implementation of this subsystem to start with the simplest policy. By merging this PR, we allow the community to effectively engage in developing this subsystem and its policies.

> I also want to see some thought put into how would this integrate with a clock managing scheme which is being talked about sometimes in arch WG

That integration would be handled by the vendor in their SoC p-state driver. The CPUFreq subsystem is simply a periodic task hinting the SoC to change its frequency; the vendor decides how to implement that. The original WG discussion on this subsystem was made with context of the current efforts in clock management and PM.

This is the introduction of a new subsystem, and I fully expect it to be iterated on by the community. Given the support and interest the RFC received in the WG and this PR, I think we are safe to proceed, as hardware implementations are already in the works and soon to follow.

seankyer commented:
Rebased and pushed further changes to documentation as well as addressed some nits.

bjarki-andreasen commented on subsys/cpu_load/Kconfig:
```diff
@@ -0,0 +1,20 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig CPU_LOAD_SUBSYS
```
SUBSYS should not be in the configs name, it doesn't really convey much information, CPU_LOAD_INFO or CPU_USAGE even would be more descriptive :)

bjarki-andreasen commented on include/zephyr/cpu_load/cpu_load.h:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @retval 0 In case of success, nonzero in case of failure.
+ */
+int cpu_load_get(int cpu_id, int *load);
```
the load arg does not need to be signed, and could be shortened to uint8_t. I would even prefer using the return value to store the load:
```
* @retval CPU load in percent if success.
* @retval -errno code in case of failure.
```

seankyer replied:
I will switch to use uint8_t. But since the implementation calls `runtime_stats` APIs, which could potentially return an error code, I'd like to stick to using a pointer.

bjarki-andreasen replied:
The return value would definitely be an int :) I think it looks good, it better for the stack (saves a pointer) and it definitely has precedence:
https://github.com/zephyrproject-rtos/zephyr/blob/72fdf7e7f3123932fff3545ccb28f42a207f18de/include/zephyr/drivers/rtc.h#L334
I don't have a strong opinion, but I have to be honest and say I would using the return value

decsny commented:
> @decsny I think you have system-level concerns of this subsystem and a concern about there being no real SoC implementation with this PR.
> 
> To address the SoC implementation concern, I should put into context that this RFC has been reviewed by the Architecture WG about a month ago. I've collaborated with community members during the WG, RFC review and this PR to produce the current implementation. There has been engagement across multiple vendors. @bjarki-andreasen has already mocked support for his Nordic SoC #93905, and I believe @peter-mitsis at Intel has as well. At ADI we are also implementing support on our SoC which we will upstream.

I want these things to actually be visible by this PR, linked somehow, with actual measured outcomes of how it is working compared to the chip intended use case, and what application they run. To me the things linked so far, looks like they dont even build. A "mock" is not a real proof that something will work, and neither is some internal "trust me bro" branch.

> I believe it's reasonable for this PR to be merged with the dummy `native_sim` driver, serving as both a CI target and a reference implementation, with support for other SoCs to be added through subsequent PRs.

I don't agree,  I think the first implementation in tree should be on a real constrained hardware with real requirements.

> I believe your system concerns are addressable with future development of this subsystem, its policies, and API.

This is making a lot of assumption. The core issue I have here is not that the subsystem is not fleshed out but that the fundamental design might be competing with what we are trying to come up with to synergize PM, clock manager, and DVFS. The clock manager is core to all of these things. Right now we are ending up with a lot of parallel state machines being defined project-specific in DT even though all of these things should be considered deeply intertwined and related. The PM has some states, and this cpufreq/DVFS will have some p-states. Right now it looks fine because mostly power modes are orthogonal to DVFS as long as we dont have a clock manager due to by nature, the cpu freq is only relevant when CPU is active, whereas PM is only relevant when CPU is inactive. But now this clock manager being proposed by @danieldegrasse and others is really the thing that will unify the whole system, and it is having yet another state machine being described parallel to both of these, and I think that may be a bad design. So the issue I have is not with your isolated design for this but with the fact that I think we have not thought about the whole system architecture very thoroughly and maybe this design is fundamentally not going to work, and we don't even know yet.

> > In a real system to change the frequency and voltage could be a very complicated thing, possibly needing to call on a lot of different software components
> 
> Correct. That complexity is real, but it doesn't invalidate having a common API/abstraction layer. Any vendor supporting Zephyr's CPUFreq subsystem must account for those details in their p-state driver and select an appropriate combination of policy and period. If an SoC cannot switch DVFS modes in a performant way, then a load-based policy may not make sense for it (and it should select another).

I am not against having a common layer for DVFS, in fact, it's something very much needed. And like I said, your proposal seems good to me _now_, but I am worried about the integration of all those things I mentioned above and want to be more patient about these things to make sure they all synergize.

> > we don't know what is the future load, ie what is the load pattern, beyond just an average. I think what you have done is provide the possibility of some backends for policies, but the current one is such an oversimplification of the real life problem.
> 
> This concern, and the concern of a circular load-loop, is a fair critique of the _on-demand_ policy, but not of the subsystem itself. If the policy is too simple, the solution is to add a better one. I see there being many policies eventually, some of which combine multiple metrics. Looking ahead, there has already been discussions in the Architecture WG to add a temperature metric and a temperature-based policy [#92123 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/92123#issuecomment-3049414846). I can also think of the rationale behind a load-trend based policy. The API could also be extended with a "force" function, letting an application request a p-state preemptively if it knows heavy processing is coming.
> The on-demand policy may indeed be too simple for some systems, but where do we start? It makes perfect sense for the initial implementation of this subsystem to start with the simplest policy. By merging this PR, we allow the community to effectively engage in developing this subsystem and its policies.

I get it, I completely understand what this PR is proposing. I know this is only one policy, but the point I am saying is that IMO, this policy is not a convincing one to say that this system will generically work for more complicated policies. That's why , I want to see a real example on a real constrained device with an application with real requirements, with a real-life policy. You got this one based on the idea from Linux, but linux-class systems have completely different requirements than the ones on Zephyr.


>
> > I also want to see some thought put into how would this integrate with a clock managing scheme which is being talked about sometimes in arch WG
> 
> That integration would be handled by the vendor in their SoC p-state driver. The CPUFreq subsystem is simply a periodic task hinting the SoC to change its frequency; the vendor decides how to implement that. The original WG discussion on this subsystem was made with context of the current efforts in clock management and PM.
> 
> This is the introduction of a new subsystem, and I fully expect it to be iterated on by the community. Given the support and interest the RFC received in the WG and this PR, I think we are safe to proceed, as hardware implementations are already in the works and soon to follow.

I don't agree with this mindset. When something like this is introduced, people pool their resources into it whether it is the right direction or not, and it will become sticky compared to working it out pre-merge. This should be common sense for many people working on open source or any software project. This is probably why, Linux does this type of work in separate branches from the mainline, and doesn't shy away from merge commits (as an aside, I do not understand why Zephyr project forbids merge commits in the mainline history, they are extremely useful AND still bisect-able)

bjarki-andreasen reviewed (dismissed):
I like CPU_LOAD_METRIC

peter-mitsis commented on subsys/cpu_freq/policies/on_demand/on_demand.c:
```diff
@@ -0,0 +1,55 @@
+ * first P-state where the cpu_load is greater than or equal to the trigger threshold
+ * of the P-state.
+ */
+int cpu_freq_policy_select_pstate(const struct pstate **pstate_out)
```
After experimenting with this in attempts to adapt it for SMP (based on IPI work queue proposal in #93795)  I think it may be worthwhile to change the API slightly to accept a CPU ID as a parameter.

int cpu_freq_policy_select_pstate(const struct pstate **pstate_out, unsigned int cpu_id);

This would necessitate that the call to cpu_load_get(0) be changed to cpu_load_get(cpu_id);

Without this, we start running into a couple of issues in SMP where if this is being called at thread level, we could find that the thread gets bounced around between CPUs and then we wind up setting the pstate for the wrong CPU.

bjarki-andreasen replied:
this would also require adding a property to the cpus themselves in devicetree, pointing to the subset of performance states each CPU supports

seankyer replied:
My original question is just if I need to update the signature of the driver API as well as the policy API so it's easier to extend for SMP later @peter-mitsis?

peter-mitsis replied:
I've been playing some more with SMP, and I completely retract my desire to add the CPU ID as a parameter.

nashif reviewed (commented):
dont think this one single function which extends and builds on top existing metric/usage code justifies a new subsystem unless I am missing something. IMO this should either be along side existing 'usage' implementation or in lib/os. 

Either way, this can be moved later, do not wish to delay this any further.

seankyer commented:
@decsny do you have any actionable edits for this PR you are requesting?

danieldegrasse reviewed (commented):
If I'm reading this right, there is essentially no user facing API added here, right? IE the user turns on the on-demand governor and then the CPU scaling happens transparently?

I ask because I am inclined to agree with @decsny  that this subsystem will likely need to be made more generic to support DVFS as a whole- however if we have no user APIs present I'm personally fine with the RFC as is. Essentially it lets us take incremental steps forward and work on a full DVFS design later.

A few things I think we should consider moving forwards though:
- Do all CPUs need to use the same governor architecture? (may matter in big/little designs)
- should we allow changing the governor at runtime?

As far as #72102, I do agree that this framework sits on top- maybe we can have a generic pstate driver once the framework is finalized, but they are still complementary.

seankyer commented:
> If I'm reading this right, there is essentially no user facing API added here, right? IE the user turns on the on-demand governor and then the CPU scaling happens transparently?
> 
> I ask because I am inclined to agree with @decsny  that this subsystem will likely need to be made more generic to support DVFS as a whole- however if we have no user APIs present I'm personally fine with the RFC as is. Essentially it lets us take incremental steps forward and work on a full DVFS design later.
> 
> A few things I think we should consider moving forwards though:
> - Do all CPUs need to use the same governor architecture? (may matter in big/little designs)
> - should we allow changing the governor at runtime?
> 
> As far as #72102, I do agree that this framework sits on top- maybe we can have a generic pstate driver once the framework is finalized, but they are still complementary.

Yes you understand correctly.

For the first point, it's a good question. I think it relates to the SMP support @peter-mitsis is developing. Not sure about the second point yet. Could be interesting to visit that possibility later.

kartben reviewed (dismissed):
Refresh +1 for docs and samples. Really nice work!

danieldegrasse commented:
> Daniel is an agreeable person and willing to work with people. And yet he said it is probably not proven enough that it will be able to do DVFS use cases.

I generally am, yes- I don't really appreciate this being used to present my passive approval as less relevant to the conversation.

> In my experience, it is WAY EASIER to fix up an unmerged thing than to undo/change/rework an existing merged thing that people already tried to start using. Why don't we consider putting this on a similar "collab branch" so that people can start feeling out how it actually works for them before we make it "sticky" on the main branch?

Is there something about the current design you see as likely to limit DVFS in the future? From what I can tell, this design allows the SOC to describe exactly what is in a pstate, so there is nothing stopping a future DVFS implementation from storing DVFS state data there. All this framework does is select an SOC-defined setting for a given load threshold.

bjarki-andreasen commented:
@decsny I approved this PR on the basis of #93905. I just now pushed to it to align it with this PR, and add it to the tests. Its not a mock, though it is indeed trivial, which I would think is a good thing. From a design perspective, I think this PR provides a good starting point. There are definitely questions regarding the on_demand policy, like will it be constantly jumping back and forth between p-states as a higher p-state theoretically lowers the CPU load given higher processing capability, so the load goes down a bit, so we go down a p-state, and repeat. You pointed to the state we would wake up into, which I thought would be the one before we went to sleep, as we can still idle while having applied a higher p-state, but that can definitely be discussed.

The only thing I see us "locking in to" is the way where we define p-states (the devicetree), the name of the subsystem, and generic API to apply a p-state. I think that is ok, and given its experimental, we can change it, I really don't believe it will be that controversial, take the stepper subsystem for example, that thing is changing a lot even though its on main, and I don't think that subsystem would be anywhere near where it is now if it was kept on a dev branch.

Regarding extending PM, I initially suggested p-states could be defined as sleep states [discord message](https://discord.com/channels/720317445772017664/733037675786797127/1379903774633754655) but with the limited scope of this PR, it only cares about the CPU speed clock speed, anything else is up to the soc, including clock management for example. But those are implementation details which should not actually affect the subsystem in this PR.

seankyer commented:
> Should this feature be turned of if CONFIG_PM=y? Otherwise, do we program into the PM subsys to update the clock freq's under certain conditions?

No, these two subsystems can run at the same time. 
Do you have a condition in mind where PM subsystem may need to update the clock? It may be something that varies system to system and should be accounted for in the pm/pstate drivers. I think the responsibilities are quite separate for now tho, PM is concerned with sleep modes, CPUFreq is concerned with DVFS.

seankyer commented:
> I think the concern he is having is similar to what I was pointing out above, which is that upon wakeup from low power, it's not clear what the contract _should_ be for which pstate to take, and may be different depending on the platform for how the hardware actually changes during low power that may not actually match anymore the pstate which was previously configured before low power. Which is one of the points of why I said, maybe it makes sense to integrate this with the PM subsystem more closely, even if the policy managers are separate.

Right, I made the assumption that the dividers for the clocks should not change between sleep mode transitions, but I can imagine an SoC existing where in order to wakeup from a sleep mode, it needs to explicitly set the clock frequency (if it's something that is not persisted between sleep modes). So, we should be aware that such an SoC could exist and this system needs to support it. 

This can be addressed by:
- Creating an API in CPUFreq to apply the latest chosen pstate. `cpu_freq_apply_last()` or something like that.
  - The CPUFreq subsystem just stores locally whatever the last applied pstate was to be recalled later
- The SoC's PM `pm_state_exit_post_ops()` implementation makes a call to that API
  - I don't think we need to hook this into the PM subsystem directly, because not all SoCs will require this.
  - If we do want to hook this into PM, we should offer a Kconfig to disable this call, for optimization sake (if the previous frequency persists through power modes).

I'd say that at this point, my opinion is that I'd like to keep the two subsystems separate for now, because I think it's simpler to work on this if it's isolated, and I also think there is a use-case for using CPUFreq, but not PM sleep-modes in a system. If they don't need to be coupled now, I'd prefer to keep them separate as long as we can. Of course, this is not a forever decision if we hit some limit in the future, this can be changed.

seankyer committed:
doc: release: Add CPUFreq to 4.3 release notes

Add CPUFreq and enabling Kconfig to 4.3 release notes.

Signed-off-by: Sean Kyer <sean.actor@gmail.com>

seankyer commented on doc/services/cpu_freq/index.rst:
```diff
@@ -0,0 +1,67 @@
+the SoC's p-state driver.
+
+With the exception of a SoC custom p-state policy, the CPU Freq standard policies shall not make use
+of any SoC specific properties.
```
Removed it

seankyer commented on doc/services/cpu_freq/index.rst:
```diff
@@ -0,0 +1,67 @@
+over the lifetime of the program. See :ref:`Kernel Timing <kernel_timing>` for more information.
+
+The CPU Frequency Scaling subsystem runs as a handler function to a ``k_timer``, which means it runs
+as an interrupt service routine (ISR). The SoC p-state driver must ensure that its implementation of
```
Thanks, updated it.

seankyer commented on doc/services/cpu_freq/index.rst:
```diff
@@ -0,0 +1,67 @@
+cannot be completed reasonably in an ISR context, it is recommended that the p-state driver of the
+SoC implements its task as a workqueue item.
+
+CPU Frequency Scaling subsystem is not compatible with SMP as of now.
```
Added a sentence explaining.

seankyer commented on doc/services/cpu_load/index.rst:
```diff
@@ -0,0 +1,16 @@
+thread and dividing it by the total cycle count of the CPU.
+
+This getter should be called on a periodic basis as to ensure that the CPU load is accurately
+represented over time.
```
Makes sense.

seankyer commented on doc/services/cpu_load/index.rst:
```diff
@@ -0,0 +1,16 @@
+CPU Load
+########
+
+The CPU Load subsystem makes use of the ``usage`` kernel feature to implement
```
Okay, removed it.

seankyer commented on dts/bindings/p_state/zephyr,native-sim-pstate.yaml:
```diff
@@ -0,0 +1,16 @@
+include: "zephyr,pstate.yaml"
+
+properties:
+  pstate-id:
+    type: int
+    required: true
+    description: |
+      Identifier of performance state.
```
Sorry, what do you mean?

decsny replied:
nevermind

seankyer commented on dts/bindings/p_state/zephyr,pstate.yaml:
```diff
@@ -0,0 +1,15 @@
+compatible: "zephyr,pstate"
+
+properties:
+  load-threshold:
+    type: int
+    description: |
+        CPU load threshold, in percent, for entering this performance state. Once
+        the CPU load exceeds this threshold, it becomes a valid performance state
+        for the P-state driver to switch to.
```
Do you have a suggestion for this? Do you think maybe it's okay to leave this the way it is for now just for the introduction?

seankyer commented on tests/subsys/cpu_freq/cpu_freq_soc/src/main.c:
```diff
@@ -0,0 +1,40 @@
+
+	LOG_INF("%d p-states defined for %s", ARRAY_SIZE(soc_pstates_dt), CONFIG_BOARD_TARGET);
+
+	zassert_equal(cpu_freq_pstate_set(NULL), -EINVAL, "Expected -EINVAL for NULL pstate");
```
I updated the `@retval` in cpu_freq.h to state it will return an `-errno` in case of failure.

seankyer commented on tests/subsys/cpu_freq/cpu_freq_soc/src/main.c:
```diff
@@ -0,0 +1,40 @@
+
+		/* Set performance state using pstate driver */
+		ret = cpu_freq_pstate_set(state);
+		zassert_equal(ret, 0, "Failed to set p-state %d", i);
```
This just tests SoC integration with the subsystem while a developer is adding support. I got inspiration from `tests/subsys/pm/power_mgmt_soc`.

seankyer commented on samples/subsys/cpu_freq/on_demand/src/main.c:
```diff
@@ -0,0 +1,74 @@
+
+	curr_work_time_ms = 0;
+
+	/* Set timer to change cpu scenario periodically */
+	k_timer_start(&timer, K_MSEC(CONFIG_CPU_FREQ_INTERVAL_MS),
+		      K_MSEC(CONFIG_CPU_FREQ_INTERVAL_MS));
```
If I sync the two to the same timer, the sample can easily be modified to see what your CPU will do at different simulated loads. 

I actually originally had it the way you suggested, but I changed it based on someone's comment previously and I think what I landed on is nice and readable. I think it's useful to keep it simple, but that's just my opinion.

decsny replied:
> I think what I landed on is nice and readable.

it's also completely unrealistic

seankyer replied:
But I thought a sample is just supposed to demonstrate the APIs and features of a subsystem. This sample is just showing how the `on_demand` policy works and how after a period of simulated load, the policy algorithm switches to a different pstate. It's not trying to pose like a system test or something.

More samples can be added to test other policies or the subsystem more generically? I like the idea for the sample you posted here https://github.com/zephyrproject-rtos/zephyr/pull/93461#discussion_r2311977154

cfriedt replied:
@decsny is this a blocking comment?

seankyer commented on include/zephyr/cpu_freq/cpu_freq.h:
```diff
@@ -0,0 +1,46 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_SUBSYS_CPU_FREQ_H_
+#define ZEPHYR_SUBSYS_CPU_FREQ_H_
```
I would like to call it CPUFreq since this is heavily inspired by that subsystem in Linux and I think people will know what it's doing when they see the name. I'm open to suggestions, but my opinion would be to leave it.

decsny replied:
that's fine not an important deal

seankyer commented on soc/native/inf_clock/cpu_freq.c:
```diff
@@ -0,0 +1,54 @@
+	LOG_DBG("Setting performance state: %d", state_id);
+
+	switch (state_id) {
+	case 0:
+		LOG_DBG("Setting P-state 0: Nominal Mode");
+		break;
+	case 1:
+		LOG_DBG("Setting P-state 1: Low Power Mode");
+		break;
+	case 2:
+		LOG_DBG("Setting P-state 2: Ultra-low Power Mode");
+		break;
```
ah yes, one requirement I've put onto the SoC integrating this subsystem is that CPUFreq updates cannot impact the Zephyr requirement aht the system timer remains constant over the lifetime of the program https://docs.zephyrproject.org/latest/kernel/services/timing/clocks.html. I recorded this in the usage considerations section of the cpu_freq/index.rst https://github.com/zephyrproject-rtos/zephyr/pull/93461/files#diff-fb62368509620540fa52de00ee5a1ad6c709c5466b7bf0f0eecab2c0c8c523f1R61

I do not think it is impossible to update frequencies at runtime and compensate the scheduling/timing, but like you said in your other comment, it would be hard. 
Maybe the Clock Manager RFC will enable SoCs that don't have HW isolation of clocking for this to still run this subsystem by compensating for the frequency change at runtime. But for now, this is a requirement, otherwise the scheduling will be messed up as soon as the system clock frequency changes.

cfriedt replied:
@decsny is this a blocking comment?

seankyer commented on subsys/cpu_freq/cpu_freq.c:
```diff
@@ -0,0 +1,54 @@
+		return;
+	}
+	LOG_DBG("Selected pstate: load_threshold=%d, config=%p", pstate_next->load_threshold,
+		pstate_next->config);
```
I agree completely. Removed it.

seankyer commented:
Latest push:
- I added `EXPERIMENTAL` to the Kconfigs
- added versioning to the public headers
- created release note for this subsystem
- addressed some reviewer feedback

seankyer commented on doc/services/cpu_freq/policies/on_demand.rst:
```diff
@@ -0,0 +1,12 @@
+The On-Demand policy will iterate through the defined P-states and select the first P-state of which the
+CPU load exceeds the defined threshold.
+
+For an example of the on-demand policy refer to the :zephyr:code-sample:`cpu_freq_on_demand` sample.
```
okay, I got you. I was confusing deadlines with general kernel scheduling. I completely agree that this algorithm is not something like just-in-time scaling, which would be a pretty cool policy to add next. I will document this.

bjarki-andreasen replied:
Regarding just in time scaling, this "policy" should be able to accomidate predictable task/event latency #83592 which covers any "thing" which affects event handling latency (sleep modes, memory access latencies, interrupt handling latencies etc.) Where each p-state would be listed in order of latency :)

seankyer commented:
Thanks everyone
